### PR TITLE
feat(refactor-db) enforce either document or collection permissions

### PR DESCRIPTION
**Pull request by kodumbeats**
## What does this PR do?

This PR adds support for separate enforcement of collection and document permissions. To do so, this PR:
- Makes the changes to Utopia to be able to skip permission validation where unneeded ([on branch](https://github.com/kodumbeats/database/tree/dev))
- Uses the `Authorization::skip()` method to skip auth when satisfying collection permissions. (https://github.com/utopia-php/database/pull/55)

The document that we pass to `Authorization::__construct()` isn't used at all in the Authorization validator. We could use this to instead pass the collection and automatically skip when needed.

## Test Plan

Tests have been added.
- [x] test listDocuments
- [ ] add more permission tests

## Related PRs and Issues

N/A

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes.

### CODE REVIEW

File: app/controllers/api/database.php [hunk 1]
```diff
@@ -1206,6 +1230,7 @@
             throw new Exception($validator->getDescription(), 400);
         }
 
+        $afterDocument = null;
```
Comment (TorstenDittmann) [inline app/controllers/api/database.php:1233]: Why this addition?

If `$afterDocument` is never assigned, `$afterDocument ?? null` will result in `null` anyway 👍🏻

File: tests/e2e/Services/Database/DatabaseBase.php [hunk 2]
```diff
@@ -1090,6 +1102,144 @@ public function testDefaultPermissions(array $data):array
...
+    {
+        $user = 'user:' . $this->getUser()['$id'];
+        $collection = $this->client->call(Client::METHOD_POST, '/database/collections', array_merge([
+            'content-type' => 'application/json',
+            'x-appwrite-project' => $this->getProject()['$id'],
+            'x-appwrite-key' => $this->getProject()['apiKey']
+        ]), [
+            'collectionId' => 'unique()',
+            'name' => 'enforceCollectionPermissions',
+            'permission' => 'collection',
```
Comment (TorstenDittmann) [inline tests/e2e/Services/Database/DatabaseBase.php:1116]: Reminder for myself to address this in the migration 👍🏻

File: app/controllers/api/database.php (hunk 1 above)
Comment (kodumbeats) [reply app/controllers/api/database.php]: Ah, you're right - I removed it in e1c7e49

**Review by lohanidamodar — APPROVED** [review]
Looks good to me!

**Review by TorstenDittmann — APPROVED** [review]

**Pull request merged**